File: Tools/Sources/StuffToolCore/IconCatalog.swift
import Foundation

public struct AppIconDescriptor: Codable, Equatable, Sendable {
    public let id: String
    public let displayName: String
    public let alternateIconName: String?
    public let previewImageName: String

    public init(
        id: String,
        displayName: String,
        alternateIconName: String?,
        previewImageName: String,
    ) {
        self.id = id
        self.displayName = displayName
        self.alternateIconName = alternateIconName
        self.previewImageName = previewImageName
    }
}

public struct AppIconManifest: Codable, Equatable, Sendable {
    public var icons: [AppIconDescriptor]

    public init(icons: [AppIconDescriptor]) {
        self.icons = icons
    }
}

public struct IconImageData: Equatable, Sendable {
    public let data: Data

    public init(data: Data, pathDescription: String) throws {
        let signature: [UInt8] = [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]
        guard data.count >= 24, Array(data.prefix(8)) == signature else {
            throw IconCatalogFailure.message("not a PNG: \(pathDescription)")
        }
        let width = Self.bigEndianUInt32(data[16 ..< 20])
        let height = Self.bigEndianUInt32(data[20 ..< 24])
        guard width == 1024, height == 1024 else {
            throw IconCatalogFailure.message(
                "\(pathDescription) is \(width)x\(height); app icons must be 1024x1024",
            )
        }
        self.data = data
    }

    private static func bigEndianUInt32(_ bytes: Data.SubSequence) -> UInt32 {
        bytes.reduce(0) { ($0 << 8) | UInt32($1) }
    }
}

public struct IconAdditionPlan: Equatable, Sendable {
    public let setName: String
    public let icon: AppIconDescriptor
    public let manifest: AppIconManifest
    public let light: IconImageData
    public let dark: IconImageData?
    public let tinted: IconImageData?

    public init(
        setName: String,
        icon: AppIconDescriptor,
        manifest: AppIconManifest,
        light: IconImageData,
        dark: IconImageData?,
        tinted: IconImageData?,
    ) {
        self.setName = setName
        self.icon = icon
        self.manifest = manifest
        self.light = light
        self.dark = dark
        self.tinted = tinted
    }
}

public struct IconRemovalPlan: Equatable, Sendable {
    public let icon: AppIconDescriptor
    public let manifest: AppIconManifest

    public init(icon: AppIconDescriptor, manifest: AppIconManifest) {
        self.icon = icon
        self.manifest = manifest
    }
}

public enum IconCatalogFailure: Error, Equatable, CustomStringConvertible, Sendable {
    case message(String)

    public var description: String {
        switch self {
            case let .message(message): message
        }
    }
}

/// Parses and calculates complete icon-catalog operations without touching disk.
public struct IconCatalogPlanner: Sendable {
    private static let primarySet = "AppIcon"
    private static let primaryID = "classic"

    public init() {}

    public func decodeManifest(_ data: Data, pathDescription: String) throws -> AppIconManifest {
        let manifest: AppIconManifest
        do {
            manifest = try JSONDecoder().decode(AppIconManifest.self, from: data)
        } catch {
            throw IconCatalogFailure.message(
                "\(pathDescription) is not valid JSON: \(error)",
            )
        }
        try validate(manifest)
        return manifest
    }

    public func addition(
        manifest: AppIconManifest,
        light: IconImageData,
        lightFilename: String,
        name explicitName: String?,
        id explicitID: String?,
        dark: IconImageData?,
        tinted: IconImageData?,
    ) throws -> IconAdditionPlan {
        let sourceStem = URL(filePath: lightFilename).deletingPathExtension().lastPathComponent
        let name = explicitName ?? Self.pascalCase(sourceStem)
        guard name.isEmpty == false else {
            throw IconCatalogFailure.message("couldn't derive a name; pass --name")
        }
        let id = explicitID ?? Self.slug(name)
        guard id.isEmpty == false else {
            throw IconCatalogFailure.message("couldn't derive an id; pass --id")
        }
        let setName = Self.primarySet + Self.pascalCase(name)
        guard setName != Self.primarySet, id != Self.primaryID else {
            throw IconCatalogFailure.message(
                "\"\(Self.primaryID)\" / \"\(Self.primarySet)\" is the reserved primary icon",
            )
        }
        guard manifest.icons.contains(where: { $0.id == id }) == false else {
            throw IconCatalogFailure.message(
                "an icon with id \"\(id)\" already exists (use --id to pick another)",
            )
        }
        guard manifest.icons.contains(where: { $0.alternateIconName == setName }) == false else {
            throw IconCatalogFailure.message("an icon named \"\(setName)\" already exists")
        }

        let icon = AppIconDescriptor(
            id: id,
            displayName: name,
            alternateIconName: setName,
            previewImageName: setName,
        )
        var updated = manifest
        updated.icons.append(icon)
        try validate(updated)
        return IconAdditionPlan(
            setName: setName,
            icon: icon,
            manifest: updated,
            light: light,
            dark: dark,
            tinted: tinted,
        )
    }

    public func removal(manifest: AppIconManifest, target: String) throws -> IconRemovalPlan {
        if [Self.primaryID, Self.primarySet.lowercased()].contains(target.lowercased()) {
            throw IconCatalogFailure.message("the primary \"Classic\" icon can't be removed")
        }
        guard let index = manifest.icons.firstIndex(where: { icon in
            [icon.id, icon.displayName, icon.alternateIconName ?? ""]
                .contains(where: { $0.lowercased() == target.lowercased() })
        }) else {
            throw IconCatalogFailure.message(
                "no icon matching \"\(target)\" (try ./icons --list)",
            )
        }
        let icon = manifest.icons[index]
        guard icon.alternateIconName != nil else {
            throw IconCatalogFailure.message("the primary \"Classic\" icon can't be removed")
        }
        var updated = manifest
        updated.icons.remove(at: index)
        try validate(updated)
        return IconRemovalPlan(icon: icon, manifest: updated)
    }

    private func validate(_ manifest: AppIconManifest) throws {
        var identifiers: Set<String> = []
        var alternateNames: Set<String> = []
        for icon in manifest.icons {
            guard icon.id.isEmpty == false else {
                throw IconCatalogFailure.message("the icon manifest contains an empty id")
            }
            guard icon.displayName.isEmpty == false else {
                throw IconCatalogFailure.message(
                    "the icon manifest contains an empty display name for \"\(icon.id)\"",
                )
            }
            guard icon.previewImageName.isEmpty == false else {
                throw IconCatalogFailure.message(
                    "the icon manifest contains an empty preview name for \"\(icon.id)\"",
                )
            }
            guard identifiers.insert(icon.id).inserted else {
                throw IconCatalogFailure.message(
                    "the icon manifest contains duplicate id \"\(icon.id)\"",
                )
            }
            if let alternate = icon.alternateIconName {
                guard alternateNames.insert(alternate).inserted else {
                    throw IconCatalogFailure.message(
                        "the icon manifest contains duplicate asset \"\(alternate)\"",
                    )
                }
                guard URL(filePath: alternate).lastPathComponent == alternate else {
                    throw IconCatalogFailure.message(
                        "the icon manifest contains invalid asset name \"\(alternate)\"",
                    )
                }
            }
        }
        guard manifest.icons.contains(where: {
            $0.id == Self.primaryID && $0.alternateIconName == nil
        }) else {
            throw IconCatalogFailure
                .message("the icon manifest is missing the Classic primary icon")
        }
    }

    private static func pascalCase(_ text: String) -> String {
        text
            .split(whereSeparator: {
                $0.isASCII == false || $0.isLetter == false && $0.isNumber == false
            })
            .map { part in
                guard let first = part.first else { return "" }
                return String(first).uppercased() + part.dropFirst()
            }
            .joined()
    }

    private static func slug(_ text: String) -> String {
        text.lowercased().filter { $0.isASCII && ($0.isLetter || $0.isNumber) }
    }
}

/// Emits the exact two-space JSON layout historically produced by `./icons`.
public struct IconCatalogRenderer: Sendable {
    public init() {}

    public func manifestData(_ manifest: AppIconManifest) throws -> Data {
        try OrderedJSON.object([
            ("icons", .array(manifest.icons.map { icon in
                .object([
                    ("id", .string(icon.id)),
                    ("displayName", .string(icon.displayName)),
                    ("alternateIconName", icon.alternateIconName.map(OrderedJSON.string) ?? .null),
                    ("previewImageName", .string(icon.previewImageName)),
                ])
            })),
        ]).data()
    }

    public func appContentsData(
        setName: String,
        hasDark: Bool,
        hasTinted: Bool,
    ) throws -> Data {
        var images: [OrderedJSON] = [
            .object([
                ("filename", .string("\(setName).png")),
                ("idiom", .string("universal")),
                ("platform", .string("ios")),
                ("size", .string("1024x1024")),
            ]),
        ]
        if hasDark {
            images.append(
                appearanceImage(
                    value: "dark",
                    filename: "\(setName)-Dark.png",
                    appIcon: true,
                ),
            )
        }
        if hasTinted {
            images.append(
                appearanceImage(
                    value: "tinted",
                    filename: "\(setName)-Tinted.png",
                    appIcon: true,
                ),
            )
        }
        return try contentsData(images: images)
    }

    public func previewContentsData(setName: String, hasDark: Bool) throws -> Data {
        var images: [OrderedJSON] = [
            .object([
                ("filename", .string("\(setName).png")),
                ("idiom", .string("universal")),
            ]),
        ]
        if hasDark {
            images.append(
                appearanceImage(
                    value: "dark",
                    filename: "\(setName)-Dark.png",
                    appIcon: false,
                ),
            )
        }
        return try contentsData(images: images)
    }

    private func contentsData(images: [OrderedJSON]) throws -> Data {
        try OrderedJSON.object([
            ("images", .array(images)),
            ("info", .object([
                ("author", .string("xcode")),
                ("version", .integer(1)),
            ])),
        ]).data()
    }

    private func appearanceImage(value: String, filename: String, appIcon: Bool) -> OrderedJSON {
        var members: [(String, OrderedJSON)] = [
            ("appearances", .array([
                .object([
                    ("appearance", .string("luminosity")),
                    ("value", .string(value)),
                ]),
            ])),
            ("filename", .string(filename)),
            ("idiom", .string("universal")),
        ]
        if appIcon {
            members += [
                ("platform", .string("ios")),
                ("size", .string("1024x1024")),
            ]
        }
        return .object(members)
    }
}

private indirect enum OrderedJSON {
    case object([(String, OrderedJSON)])
    case array([OrderedJSON])
    case string(String)
    case integer(Int)
    case null

    func data() throws -> Data {
        var output = ""
        try render(indent: 0, into: &output)
        output += "\n"
        return Data(output.utf8)
    }

    private func render(indent: Int, into output: inout String) throws {
        switch self {
            case let .object(members):
                guard members.isEmpty == false else {
                    output += "{}"
                    return
                }
                output += "{\n"
                for (index, member) in members.enumerated() {
                    output += Self.padding(indent + 1)
                    output += try Self.fragment(member.0)
                    output += " : "
                    try member.1.render(indent: indent + 1, into: &output)
                    output += index == members.count - 1 ? "\n" : ",\n"
                }
                output += Self.padding(indent) + "}"
            case let .array(values):
                guard values.isEmpty == false else {
                    output += "[]"
                    return
                }
                output += "[\n"
                for (index, value) in values.enumerated() {
                    output += Self.padding(indent + 1)
                    try value.render(indent: indent + 1, into: &output)
                    output += index == values.count - 1 ? "\n" : ",\n"
                }
                output += Self.padding(indent) + "]"
            case let .string(value):
                output += try Self.fragment(value)
            case let .integer(value):
                output += String(value)
            case .null:
                output += "null"
        }
    }

    private static func fragment(_ string: String) throws -> String {
        let data = try JSONSerialization.data(
            withJSONObject: [string],
            options: [.withoutEscapingSlashes],
        )
        return String(decoding: data.dropFirst().dropLast(), as: UTF8.self)
    }

    private static func padding(_ indent: Int) -> String {
        String(repeating: "  ", count: indent)
    }
}

File: Tools/icon_catalog.rb
# frozen_string_literal: true

require "fileutils"
require "json"
require "tmpdir"

require_relative "file_transaction"

# Plans and transactionally commits Where alternate-icon catalog mutations.
class IconCatalog
  class Error < StandardError; end

  PRIMARY_SET = "AppIcon"
  PRIMARY_ID = "classic"

  def initialize(root:, app_catalog:, preview_catalog:, manifest:, transaction_factory:)
    @root = File.expand_path(root)
    @app_catalog = File.expand_path(app_catalog, @root)
    @preview_catalog = File.expand_path(preview_catalog, @root)
    @manifest = File.expand_path(manifest, @root)
    @transaction_factory = transaction_factory
  end

  def list(output: $stdout)
    icons = load_manifest.fetch("icons")
    if icons.empty?
      output.puts "(no icons in the manifest)"
      return
    end

    id_width = icons.map { |icon| icon.fetch("id").length }.max
    name_width = icons.map { |icon| icon.fetch("displayName").length }.max
    icons.each do |icon|
      alternate = icon["alternateIconName"] || "(primary)"
      output.printf("  %-#{id_width}s  %-#{name_width}s  %s\n", icon.fetch("id"), icon.fetch("displayName"), alternate)
    end
  end

  def add(light:, name:, id:, dark:, tinted:, dry_run:)
    light = resolve_input(light)
    dark = resolve_optional_input(dark)
    tinted = resolve_optional_input(tinted)
    name = name.empty? ? pascal_case(File.basename(light, File.extname(light))) : name
    raise Error, "couldn't derive a name; pass --name" if name.empty?

    id = id.empty? ? slug(name) : id
    raise Error, "couldn't derive an id; pass --id" if id.empty?

    set_name = PRIMARY_SET + pascal_case(name)
    if set_name == PRIMARY_SET || id == PRIMARY_ID
      raise Error, %("#{PRIMARY_ID}" / "#{PRIMARY_SET}" is the reserved primary icon)
    end

    [light, dark, tinted].compact.each { |path| require_1024(path) }
    data = load_manifest
    icons = data.fetch("icons")
    raise Error, %(an icon with id "#{id}" already exists (use --id to pick another)) if icons.any? { |icon| icon.fetch("id") == id }
    if icons.any? { |icon| icon["alternateIconName"] == set_name }
      raise Error, %(an icon named "#{set_name}" already exists)
    end

    app_target = File.join(@app_catalog, "#{set_name}.appiconset")
    preview_target = File.join(@preview_catalog, "#{set_name}.imageset")
    [app_target, preview_target].each do |path|
      raise Error, "#{relative(path)} already exists" if File.exist?(path) || File.symlink?(path)
    end

    data["icons"] = icons + [{
      "id" => id,
      "displayName" => name,
      "alternateIconName" => set_name,
      "previewImageName" => set_name,
    }]
    validate_manifest(data)

    Dir.mktmpdir(".icons-stage-", @root) do |stage|
      staged_app = File.join(stage, "#{set_name}.appiconset")
      staged_preview = File.join(stage, "#{set_name}.imageset")
      staged_manifest = File.join(stage, "AppIcons.json")
      stage_app_icon(staged_app, set_name: set_name, light: light, dark: dark, tinted: tinted)
      stage_preview(staged_preview, set_name: set_name, light: light, dark: dark)
      write_json(staged_manifest, data)
      validate_staged_add(staged_app, staged_preview, staged_manifest, set_name)

      unless dry_run
        transaction = @transaction_factory.call
        transaction.replace(target: app_target, staged: staged_app)
        transaction.replace(target: preview_target, staged: staged_preview)
        transaction.replace(target: @manifest, staged: staged_manifest)
        transaction.commit
      end
    end

    verb = dry_run ? "Would add" : "Added"
    %(#{verb} "#{name}" (id: #{id}, asset: #{set_name}).)
  end

  def remove(target:, dry_run:)
    raise Error, %(the primary "Classic" icon can't be removed) if [PRIMARY_ID, PRIMARY_SET.downcase].include?(target.downcase)

    data = load_manifest
    icons = data.fetch("icons")
    match = icons.find do |icon|
      values = [icon.fetch("id"), icon.fetch("displayName"), icon["alternateIconName"]].compact
      values.any? { |value| value.downcase == target.downcase }
    end
    raise Error, %(no icon matching "#{target}" (try ./icons --list)) unless match
    raise Error, %(the primary "Classic" icon can't be removed) unless match["alternateIconName"]

    set_name = match.fetch("alternateIconName")
    data["icons"] = icons.reject { |icon| icon.equal?(match) }
    validate_manifest(data)
    app_target = File.join(@app_catalog, "#{set_name}.appiconset")
    preview_target = File.join(@preview_catalog, "#{set_name}.imageset")

    Dir.mktmpdir(".icons-stage-", @root) do |stage|
      staged_manifest = File.join(stage, "AppIcons.json")
      write_json(staged_manifest, data)
      validate_manifest(JSON.parse(File.read(staged_manifest)))

      unless dry_run
        transaction = @transaction_factory.call
        transaction.remove(target: app_target)
        transaction.remove(target: preview_target)
        transaction.replace(target: @manifest, staged: staged_manifest)
        transaction.commit
      end
    end

    verb = dry_run ? "Would remove" : "Removed"
    %(#{verb} "#{match.fetch('displayName')}" (id: #{match.fetch('id')}).)
  end

  private

  def require_layout
    missing = [@manifest, @app_catalog, @preview_catalog].reject { |path| File.exist?(path) }
    raise Error, "couldn't find #{missing.map { |path| relative(path) }.join(', ')} — run ./icons from the repo root" if missing.any?
  end

  def load_manifest
    require_layout
    data = JSON.parse(File.read(@manifest))
    validate_manifest(data)
    data
  rescue JSON::ParserError => error
    raise Error, "#{relative(@manifest)} is not valid JSON: #{error.message}"
  end

  def validate_manifest(data)
    raise Error, "#{relative(@manifest)} must contain an icons array" unless data.is_a?(Hash) && data["icons"].is_a?(Array)

    ids = []
    alternates = []
    data.fetch("icons").each_with_index do |icon, index|
      raise Error, "icon #{index + 1} must be an object" unless icon.is_a?(Hash)
      %w[id displayName previewImageName].each do |key|
        raise Error, "icon #{index + 1} has no #{key}" unless icon[key].is_a?(String) && !icon[key].empty?
      end
      alternate = icon["alternateIconName"]
      raise Error, "icon #{index + 1} has an invalid alternateIconName" unless alternate.nil? || (alternate.is_a?(String) && !alternate.empty?)
      ids << icon.fetch("id")
      alternates << alternate if alternate
    end
    raise Error, "manifest contains duplicate icon ids" unless ids.uniq.length == ids.length
    raise Error, "manifest contains duplicate alternate icon names" unless alternates.uniq.length == alternates.length
  end

  def stage_app_icon(directory, set_name:, light:, dark:, tinted:)
    FileUtils.mkdir_p(directory)
    images = [{ "filename" => "#{set_name}.png", "idiom" => "universal", "platform" => "ios", "size" => "1024x1024" }]
    copy(light, directory, "#{set_name}.png")
    if dark
      images << appearance_image("#{set_name}-Dark.png", "dark", platform: true)
      copy(dark, directory, "#{set_name}-Dark.png")
    end
    if tinted
      images << appearance_image("#{set_name}-Tinted.png", "tinted", platform: true)
      copy(tinted, directory, "#{set_name}-Tinted.png")
    end
    write_json(File.join(directory, "Contents.json"), "images" => images, "info" => xcode_info)
  end

  def stage_preview(directory, set_name:, light:, dark:)
    FileUtils.mkdir_p(directory)
    images = [{ "filename" => "#{set_name}.png", "idiom" => "universal" }]
    copy(light, directory, "#{set_name}.png")
    if dark
      images << appearance_image("#{set_name}-Dark.png", "dark", platform: false)
      copy(dark, directory, "#{set_name}-Dark.png")
    end
    write_json(File.join(directory, "Contents.json"), "images" => images, "info" => xcode_info)
  end

  def appearance_image(filename, value, platform:)
    image = {
      "appearances" => [{ "appearance" => "luminosity", "value" => value }],
      "filename" => filename,
      "idiom" => "universal",
    }
    image["platform"] = "ios" if platform
    image["size"] = "1024x1024" if platform
    image
  end

  def validate_staged_add(app, preview, manifest, set_name)
    app_contents = JSON.parse(File.read(File.join(app, "Contents.json")))
    preview_contents = JSON.parse(File.read(File.join(preview, "Contents.json")))
    [app_contents, preview_contents].each do |contents|
      contents.fetch("images").each do |image|
        path = File.join(contents.equal?(app_contents) ? app : preview, image.fetch("filename"))
        raise Error, "staged catalog is missing #{File.basename(path)}" unless File.file?(path)
      end
    end
    staged_data = JSON.parse(File.read(manifest))
    validate_manifest(staged_data)
    unless staged_data.fetch("icons").any? { |icon| icon["alternateIconName"] == set_name }
      raise Error, "staged manifest does not include #{set_name}"
    end
  rescue JSON::ParserError, KeyError => error
    raise Error, "staged icon output is invalid: #{error.message}"
  end

  def require_1024(path)
    raise Error, "no such file: #{relative(path)}" unless File.file?(path)

    header = File.binread(path, 24)
    raise Error, "not a PNG: #{relative(path)}" unless header.start_with?("\x89PNG\r\n\x1A\n".b)

    width, height = header.byteslice(16, 8).unpack("NN")
    raise Error, "#{relative(path)} is #{width}x#{height}; app icons must be 1024x1024" unless [width, height] == [1024, 1024]
  rescue EOFError
    raise Error, "not a PNG: #{relative(path)}"
  end

  def resolve_input(path)
    File.expand_path(path, @root)
  end

  def resolve_optional_input(path)
    path.empty? ? nil : resolve_input(path)
  end

  def copy(source, directory, filename)
    FileUtils.copy_file(source, File.join(directory, filename))
  end

  def write_json(path, data)
    output = JSON.pretty_generate(
      data,
      indent: "  ",
      space: " ",
      space_before: " ",
      object_nl: "\n",
      array_nl: "\n",
    )
    File.write(path, "#{output}\n")
  end

  def xcode_info
    { "author" => "xcode", "version" => 1 }
  end

  def pascal_case(text)
    text.split(/[^A-Za-z0-9]+/).reject(&:empty?).map { |part| part[0].upcase + part[1..] }.join
  end

  def slug(text)
    text.downcase.gsub(/[^a-z0-9]+/, "")
  end

  def relative(path)
    path.delete_prefix("#{@root}/")
  end
end

def icon_catalog_main
  catalog = IconCatalog.new(
    root: ENV.fetch("ROOT"),
    app_catalog: ENV.fetch("APP_CATALOG"),
    preview_catalog: ENV.fetch("PREVIEW_CATALOG"),
    manifest: ENV.fetch("MANIFEST"),
    transaction_factory: -> { FileTransaction.new },
  )
  dry_run = ENV.fetch("DRY_RUN") == "true"
  case ENV.fetch("MODE")
  when "list"
    catalog.list
  when "add"
    puts catalog.add(
      light: ENV.fetch("LIGHT"),
      name: ENV.fetch("NAME"),
      id: ENV.fetch("ID"),
      dark: ENV.fetch("DARK"),
      tinted: ENV.fetch("TINTED"),
      dry_run: dry_run,
    )
    puts dry_run ? "Dry run — nothing was changed." : "Run `./ide --no-open` to regenerate so the new icon is compiled in."
  when "remove"
    puts catalog.remove(target: ENV.fetch("TARGET"), dry_run: dry_run)
    unless dry_run
      puts "If it was the active icon, the app falls back to Classic on next launch."
      puts "Run `./ide --no-open` to regenerate."
    end
    puts "Dry run — nothing was changed." if dry_run
  else
    raise IconCatalog::Error, "unknown mode: #{ENV.fetch('MODE')}"
  end
  0
rescue IconCatalog::Error, FileTransaction::CleanupError, SystemCallError => error
  warn "error: #{error.message}"
  1
end

exit(icon_catalog_main) if __FILE__ == $PROGRAM_NAME
